src/main/java/org/jooq/lambda/concurrent/SameExecutorCompletionStage.java

```diff
@@ -0,0 +1,222 @@
+package org.jooq.lambda.concurrent;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+final class SameExecutorCompletionStage<T> implements CompletionStage<T> {
+    private final CompletionStage<T> delegate;
+    private final Executor defaultExecutor;
```

Comment:
```
CompletionStage<String> stage = Async.supplyAsync(() -> "test", executorA);
stage.thenApplyAsync(fn, null);
```

In this scenario I see three options
1. We favor the last used `Executor` (executorA)
2. Anytime a null `Executor` is passed we switch to the common pool.
3. Throw an exception when the pool is null. `CompletableFuture` does this already so there would be no code change and it would follow `CompletableFuture`'s convention.

```
    public <U> CompletableFuture<U> thenApplyAsync(
        Function<? super T,? extends U> fn, Executor executor) {
        return uniApplyStage(screenExecutor(executor), fn);
    }

    ...

    static Executor screenExecutor(Executor e) {
        if (!useCommonPool && e == ForkJoinPool.commonPool())
            return asyncPool;
        if (e == null) throw new NullPointerException();
        return e;
    }
```

I prefer option 3 but not very strongly, any of them work. If we pick option 1 or 2 it just needs to be documented.